src/key_values/value/any.rs

```diff
@@ -0,0 +1,97 @@
+use std::{fmt, mem};
+use std::marker::PhantomData;
+
+use super::{ToValue, KeyValueError};
+use super::backend::Backend;
+
+/// A function for converting some type `T` into a [`Value`](struct.Value.html).
+pub type FromAnyFn<T> = fn(FromAny, &T) -> Result<(), KeyValueError>;
+
+/// A helper for converting any type into a [`Value`](struct.Value.html).
+pub struct FromAny<'a>(&'a mut Backend);
+
+impl<'a> FromAny<'a> {
+    pub(super) fn value<T>(self, v: T) -> Result<(), KeyValueError>
+    where
+        T: ToValue,
+    {
+        v.to_value().inner.visit(self.0)
+    }
+
+    /// Convert a formattable type into a value.
+    pub fn debug<T>(self, v: T) -> Result<(), KeyValueError>
+    where
+        T: fmt::Debug
+    {
+        self.0.fmt(format_args!("{:?}", v))
+    }
+
+    /// Convert a `u64` into a value.
+    pub fn u64(self, v: u64) -> Result<(), KeyValueError> {
+        self.0.u64(v)
+    }
+
+    /// Convert a `i64` into a value.
+    pub fn i64(self, v: i64) -> Result<(), KeyValueError> {
+        self.0.i64(v)
+    }
+    
+    /// Convert a `f64` into a value.
+    pub fn f64(self, v: f64) -> Result<(), KeyValueError> {
+        self.0.f64(v)
+    }
+
+    /// Convert a `bool` into a value.
+    pub fn bool(self, v: bool) -> Result<(), KeyValueError> {
+        self.0.bool(v)
+    }
+
+    /// Convert a `char` into a value.
+    pub fn char(self, v: char) -> Result<(), KeyValueError> {
+        self.0.char(v)
+    }
+
+    /// Convert an empty type into a value.
+    pub fn none(self) -> Result<(), KeyValueError> {
+        self.0.none()
+    }
+
+    /// Convert a string into a value.
+    pub fn str(self, v: &str) -> Result<(), KeyValueError> {
+        self.0.str(v)
+    }
+}
+
+// `Any<'a>` is very similar to `std::fmt::Arguments<'a>`
+// It takes a &T and fn pointer and stores them in an erased structure.
+// It's a bit like an ad-hoc trait object that can accept any arbitrary
+// value without those values needing to implement any traits.
+
+#[derive(Clone, Copy)]
+pub(super) struct Any<'a> {
+    data: &'a Void,
+    from: FromAnyFn<Void>,
+}
+
+// FIXME: This would be more correct as an extern type
+// Replace once the `extern_types` feature is stable
+// and available
+struct Void {
+    _priv: (),
+    _oibit_remover: PhantomData<*mut Fn()>,
+}
+
+impl<'a> Any<'a> {
+    pub(super) fn new<T>(data: &'a T, from: FromAnyFn<T>) -> Self {
+        unsafe {
+            Any {
+                data: mem::transmute::<&'a T, &'a Void>(data),
+                from: mem::transmute::<FromAnyFn<T>, FromAnyFn<Void>>(from),
```

Comment:
This seems sketchy as the Rust calling convention for those two function types might be arbitrarily different. I think this is in the realm of "probably works but who knows". It would be good to have a compiler person take a look.

- `fn(FromAny, &T) -> Result<(), KeyValueError>`
- `fn(FromAny, &Void) -> Result<(), KeyValueError>`

Making them something like `unsafe extern "C" fn(FromAny, *const T) -> Result<(), KeyValueError>` seems like it would be sufficient to avoid undefined behavior but I don't know how much worse that makes the rest of the API.

Reply:
Yeh, this was the weird API I was hoping to get some more eyes on at the All Hands earlier this year.

It's the same approach that [`std::fmt::Argument` uses](https://github.com/rust-lang/rust/blob/master/src/libcore/fmt/mod.rs#L240-L267), but that's unstable. It's also similar to the `std::task::RawWakerVTable` which is up for stabilization now,  but includes some of the extra  type safety @eddyb suggested [in the stabilization PR](https://github.com/rust-lang/rust/pull/59739#discussion_r272793688). The impression I've gotten is that this approach is _ok_, but is well off the beaten track and not something you should reach for if you don't have to.

As for why we need it, there's a case where you're bridging between  `log` and another framework  where you effectively need:

```
&'kv T ->  Value<'kv> where T: ToValue || Debug
```

If your bridging crate doesn't own  the type `T` and it doesn't satisfy either of those bounds on its own then you're stuck. You can't wrap `&'kv T` in a newtype because you need the data within `Value` to live for `'kv`. That's where `Value::from_any` comes  in. It lets you create a `Value<'kv>` from any arbitrary `&'kv T`, regardless of the traits `T` implements.

Additionally, when we come to integrating other serialization frameworks like `serde` into `log`,  we can do it by erasing into the same `Any` mechanism, so there's no matching required.

This approach comes with drawbacks though:

- It _is_ kinda sketchy and will probably surprise any future contributors that come across it.
- Once  we make `from_any` public, I can't think of any alternative internal implementation of `Value` that could support that `&'kv T -> Value<'kv>` method.
- Supporting owned values in the future using this erased approach is much trickier.

This PR doesn't make the `from_any` method public though, so right now there's no reason to use this erased approach over an internal trait object. I just copied it from my reference implementation. For implementation simplicity I'd be on board with replacing this with the more obvious trait-object-based implementation.

Reply:
Alrighty, in https://github.com/rust-lang-nursery/log/pull/324/commits/ee05032759cf2fd6c04b6501b8c35b8e2e3942a7 I've refactored this to remove this `FromAny` infrastructure and use a pair of internal traits `Visit` and `Visitor` to preserve the structure of primitive values.